File: src/infra/instance/instanceHttpRequestStatisticsService.ts
import { actuatorClientStore } from '../actuator/actuatorClientStore';
import NodeCache from 'node-cache';
import { HasAbility } from '../utils/hasAbility';
import {
  HttpMethod,
  InstanceHttpRequestStatisticsForUriOptions,
  InstanceHttpRequestStatistics,
  InstanceHttpRequestStatisticsByMethod,
  HttpRequestOutcome,
  InstanceHttpRequestStatisticsByOutcome,
  InstanceHttpRequestStatisticsByStatus,
  InstanceHttpRequestStatisticsByException,
} from './models/httpRequestStatistics';
import { systemEvents } from '../events';

class InstanceHttpRequestStatisticsService {
  private readonly metricName = 'http.server.requests';

  private readonly statisticsCache = new NodeCache({
    stdTTL: 600,
  });

  constructor() {
    systemEvents.on('instance-updated', (instance) => {
      this.statisticsCache.del(instance.id);
    });
    systemEvents.on('instance-deleted', (instance) => {
      this.statisticsCache.del(instance.id);
    });
  }

  @HasAbility('http-request-statistics')
  async getStatistics(instanceId: string): Promise<InstanceHttpRequestStatistics[]> {
    const cached = this.statisticsCache.get(instanceId);
    if (cached) {
      return cached as InstanceHttpRequestStatistics[];
    }
    const statistics = await this.computeInstanceStatistics(instanceId);
    this.statisticsCache.set(instanceId, statistics);
    return statistics;
  }

  @HasAbility('http-request-statistics')
  async getStatisticsForUri(
    instanceId: string,
    uri: string,
    options: InstanceHttpRequestStatisticsForUriOptions
  ): Promise<InstanceHttpRequestStatistics> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName, { uri, ...options });

    return {
      uri,
      count: metric.measurements.find((measurement) => measurement.statistic === 'COUNT')?.value ?? -1,
      totalTime: metric.measurements.find((measurement) => measurement.statistic === 'TOTAL_TIME')?.value ?? -1,
      max: metric.measurements.find((measurement) => measurement.statistic === 'MAX')?.value ?? -1,
    };
  }

  @HasAbility('http-request-statistics')
  async getStatisticsForUriByMethods(
    instanceId: string,
    uri: string
  ): Promise<InstanceHttpRequestStatisticsByMethod[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName, { uri });
    const availableMethods: HttpMethod[] =
      (metric.availableTags.find((tag) => tag.tag === 'method')?.values as HttpMethod[]) ?? [];
    if (!availableMethods.length) {
      return [];
    }
    const promises = availableMethods.map(async (method) => {
      return { method, statistics: await this.getStatisticsForUri(instanceId, uri, { method }) };
    });
    const statistics = await Promise.all(promises);
    return statistics;
  }

  @HasAbility('http-request-statistics')
  async getStatisticsForUriByOutcomes(
    instanceId: string,
    uri: string
  ): Promise<InstanceHttpRequestStatisticsByOutcome[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName, { uri });
    const availableOutcomes: HttpRequestOutcome[] =
      (metric.availableTags.find((tag) => tag.tag === 'outcome')?.values as HttpRequestOutcome[]) ?? [];
    if (!availableOutcomes.length) {
      return [];
    }
    const promises = availableOutcomes.map(async (outcome) => {
      return { outcome, statistics: await this.getStatisticsForUri(instanceId, uri, { outcome }) };
    });
    const statistics = await Promise.all(promises);
    return statistics;
  }

  @HasAbility('http-request-statistics')
  async getStatisticsForUriByStatuses(
    instanceId: string,
    uri: string
  ): Promise<InstanceHttpRequestStatisticsByStatus[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName, { uri });
    const availableStatuses: string[] =
      (metric.availableTags.find((tag) => tag.tag === 'status')?.values as string[]) ?? [];
    if (!availableStatuses.length) {
      return [];
    }
    const promises = availableStatuses.map(async (status) => {
      return {
        status: Number(status) ?? -1,
        statistics: await this.getStatisticsForUri(instanceId, uri, { status: Number(status) ?? -1 }),
      };
    });
    const statistics = await Promise.all(promises);
    return statistics;
  }

  @HasAbility('http-request-statistics')
  async getStatisticsForUriByExceptions(
    instanceId: string,
    uri: string
  ): Promise<InstanceHttpRequestStatisticsByException[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName, { uri });
    const availableExceptions: string[] =
      (metric.availableTags.find((tag) => tag.tag === 'exception')?.values as string[]) ?? [];
    if (!availableExceptions.length) {
      return [];
    }
    const promises = availableExceptions.map(async (exception) => {
      return {
        exception,
        statistics: await this.getStatisticsForUri(instanceId, uri, { exception }),
      };
    });
    const statistics = await Promise.all(promises);
    return statistics;
  }

  private async computeInstanceStatistics(instanceId: string): Promise<InstanceHttpRequestStatistics[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const availableUris = await this.getAvailableUris(instanceId);
    const promises = availableUris.map(async (uri) => {
      const metric = await client.metric(this.metricName, { uri });
      const count = metric.measurements.find((measurement) => measurement.statistic === 'COUNT')?.value ?? -1;
      const totalTime = metric.measurements.find((measurement) => measurement.statistic === 'TOTAL_TIME')?.value ?? -1;
      const max = metric.measurements.find((measurement) => measurement.statistic === 'MAX')?.value ?? -1;
      return <InstanceHttpRequestStatistics>{
        uri,
        count,
        totalTime,
        max,
      };
    });

    const statistics: InstanceHttpRequestStatistics[] = await Promise.all(promises);
    return statistics;
  }

  private async getAvailableUris(instanceId: string): Promise<string[]> {
    const client = actuatorClientStore.getActuatorClient(instanceId);
    const metric = await client.metric(this.metricName);
    const uriTag = metric.availableTags.find((tag) => tag.tag === 'uri');
    if (!uriTag) {
      return [];
    }
    return uriTag.values;
  }
}

export const instanceHttpRequestStatisticsService = new InstanceHttpRequestStatisticsService();

File: daemon/src/main/kotlin/dev/krud/boost/daemon/configuration/instance/httprequeststatistics/InstanceHttpRequestStatisticsService.kt
package dev.krud.boost.daemon.configuration.instance.httprequeststatistics

import dev.krud.boost.daemon.actuator.ActuatorHttpClient
import dev.krud.boost.daemon.configuration.instance.InstanceActuatorClientProvider
import dev.krud.boost.daemon.configuration.instance.InstanceService
import dev.krud.boost.daemon.configuration.instance.ability.InstanceAbilityService
import dev.krud.boost.daemon.configuration.instance.enums.InstanceAbility
import dev.krud.boost.daemon.configuration.instance.httprequeststatistics.enums.HttpMethod
import dev.krud.boost.daemon.configuration.instance.httprequeststatistics.ro.InstanceHttpRequestStatisticsRO
import dev.krud.boost.daemon.configuration.instance.messaging.InstanceCreatedEventMessage
import dev.krud.boost.daemon.configuration.instance.messaging.InstanceDeletedEventMessage
import dev.krud.boost.daemon.configuration.instance.messaging.InstanceUpdatedEventMessage
import dev.krud.boost.daemon.utils.resolve
import io.github.oshai.KotlinLogging
import org.springframework.beans.factory.DisposableBean
import org.springframework.cache.CacheManager
import org.springframework.cache.annotation.Cacheable
import org.springframework.integration.annotation.ServiceActivator
import org.springframework.messaging.Message
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.Executors

@Service
class InstanceHttpRequestStatisticsService(
    private val instanceService: InstanceService,
    private val actuatorClientProvider: InstanceActuatorClientProvider,
    private val instanceAbilityService: InstanceAbilityService,
    cacheManager: CacheManager
) : DisposableBean {
    private val httpRequestStatisticsCache by cacheManager.resolve()
    private val executor = Executors.newCachedThreadPool()

    override fun destroy() {
        executor.shutdown()
    }

    @Cacheable(cacheNames = ["httpRequestStatisticsCache"], key = "'all_uris_' + #instanceId")
    fun getStatistics(instanceId: UUID): List<InstanceHttpRequestStatisticsRO> {
        log.debug { "Get Http Request Statistics for instance $instanceId" }
        val instance = instanceService.getInstanceFromCacheOrThrow(instanceId)
        instanceAbilityService.hasAbilityOrThrow(instance, InstanceAbility.HTTP_REQUEST_STATISTICS)
        val client = actuatorClientProvider.provide(instance)
        val availableUris = client.getAvailableUris()
        if (availableUris.isEmpty()) {
            return emptyList()
        }
        val futures = availableUris.map { uri ->
            executor.submit<InstanceHttpRequestStatisticsRO> {
                client.getStatisticsForUri(uri)
            }
        }
        return futures.map { it.get() }
    }

    @Cacheable(cacheNames = ["httpRequestStatisticsCache"], key = "'by_uri_and_method_' + #instanceId + '_' + #uri")
    fun getStatisticsByUriAndMethod(instanceId: UUID, uri: String): Map<HttpMethod, InstanceHttpRequestStatisticsRO> {
        return getStatisticsByTag(instanceId, uri, "method") { availableTagValue -> HttpMethod.valueOf(availableTagValue) }
    }

    @Cacheable(cacheNames = ["httpRequestStatisticsCache"], key = "'by_uri_and_outcome_' + #instanceId + '_' + #uri")
    fun getStatisticsByUriAndOutcome(instanceId: UUID, uri: String): Map<String, InstanceHttpRequestStatisticsRO> {
        return getStatisticsByTag(instanceId, uri, "outcome") { availableTagValue -> availableTagValue }
    }

    @Cacheable(cacheNames = ["httpRequestStatisticsCache"], key = "'by_uri_and_status_' + #instanceId + '_' + #uri")
    fun getStatisticsByUriAndStatus(instanceId: UUID, uri: String): Map<String, InstanceHttpRequestStatisticsRO> {
        return getStatisticsByTag(instanceId, uri, "status") { availableTagValue -> availableTagValue }
    }

    @Cacheable(cacheNames = ["httpRequestStatisticsCache"], key = "'by_uri_and_exception_' + #instanceId + '_' + #uri")
    fun getStatisticsByUriAndException(instanceId: UUID, uri: String): Map<String, InstanceHttpRequestStatisticsRO> {
        return getStatisticsByTag(instanceId, uri, "exception") { availableTagValue -> availableTagValue }
    }

    private fun <KeyType : Any> getStatisticsByTag(instanceId: UUID, uri: String, tag: String, keySupplier: (String) -> KeyType): Map<KeyType, InstanceHttpRequestStatisticsRO> {
        log.debug { "Get Http Request Statistics grouped by uri and tag $tag for instance $instanceId and uri $uri" }
        val instance = instanceService.getInstanceFromCacheOrThrow(instanceId)
        instanceAbilityService.hasAbilityOrThrow(instance, InstanceAbility.HTTP_REQUEST_STATISTICS)
        val client = actuatorClientProvider.provide(instance)
        val metric = client.metric(METRIC_NAME, mapOf("uri" to uri))
        val availableTagValues = metric
            .getOrThrow()
            .availableTags
            .find { it.tag == tag }
            ?.values ?: emptyList()
        val futures = availableTagValues.map { availableTagValue ->
            executor.submit<Pair<KeyType, InstanceHttpRequestStatisticsRO>> {
                keySupplier(availableTagValue) to client.getStatisticsForUri(uri, mapOf(tag to availableTagValue))
            }
        }
        return futures.associate { it.get() }
    }

    @ServiceActivator(inputChannel = "systemEventsChannel")
    protected fun onInstanceEvent(event: Message<*>) {
        when (event) {
            is InstanceCreatedEventMessage -> {
                getCacheKeys(event.payload.instanceId).forEach {
                    httpRequestStatisticsCache.evict(it)
                }
            }
            is InstanceUpdatedEventMessage -> {
                getCacheKeys(event.payload.instanceId).forEach {
                    httpRequestStatisticsCache.evict(it)
                }
            }
            is InstanceDeletedEventMessage -> {
                getCacheKeys(event.payload.instanceId).forEach {
                    httpRequestStatisticsCache.evict(it)
                }
            }
        }
    }

    private fun ActuatorHttpClient.getAvailableUris(): List<String> {
        val metric = metric(METRIC_NAME)
        return metric
            .getOrThrow()
            .availableTags
            .find { it.tag == "uri" }
            ?.values ?: emptyList()
    }

    private fun ActuatorHttpClient.getStatisticsForUri(uri: String, tags: Map<String, String> = emptyMap()): InstanceHttpRequestStatisticsRO {
        log.debug { "Get http statistics for uri $uri" }
        val metricResponse = this.metric(METRIC_NAME, mapOf("uri" to uri) + tags)
            .getOrThrow()
        val count = metricResponse.measurements.find { it.statistic == "COUNT" }?.value ?: -1.0
        val max = metricResponse.measurements.find { it.statistic == "MAX" }?.value ?: -1.0
        val totalTime = metricResponse.measurements.find { it.statistic == "TOTAL_TIME" }?.value ?: -1.0
        return InstanceHttpRequestStatisticsRO(
            uri = uri,
            count = count.toLong(),
            max = max.toLong(),
            totalTime = totalTime.toLong()
        )
    }

    companion object {
        private const val METRIC_NAME = "http.server.requests"
        private fun getCacheKeys(instanceId: UUID): Set<String> {
            return setOf(
                "all_uris_$instanceId",
                "by_uri_and_method_$instanceId",
                "by_uri_and_outcome_$instanceId",
                "by_uri_and_status_$instanceId",
                "by_uri_and_exception_$instanceId"
            )
        }
        private val log = KotlinLogging.logger { }
    }
}